developer/src/kmldmlc/src/keyman/compiler/errors.ts

```diff
@@ -9,19 +9,44 @@ export enum CompilerErrorSeverity {
   Error_Mask =    0x0FFFFF,
 };
 
-export enum CompilerErrors {
-  ERROR_InvalidNormalization = CompilerErrorSeverity.Error | 0x0001, // `Invalid normalization form '${}'`
-  ERROR_InvalidLocale        = CompilerErrorSeverity.Error | 0x0002, // `Invalid BCP 47 locale form '${}'`
-};
+const m = (code: number, message: string) => { return { code, message } };
+// const SevInfo = CompilerErrorSeverity.Info;
+const SevHint = CompilerErrorSeverity.Hint;
+// const SevWarn = CompilerErrorSeverity.Warn;
+const SevError = CompilerErrorSeverity.Error;
+// const SevFatal = CompilerErrorSeverity.Fatal;
+
+export class CompilerErrors {
+  static InvalidNormalization = (o:{form: string}) => m(this.ERROR_InvalidNormalization, `Invalid normalization form '${o.form}`);
+  static ERROR_InvalidNormalization = SevError | 0x0001;
+
+  static InvalidLocale = (o:{tag: string}) => m(this.ERROR_InvalidLocale, `Invalid BCP 47 locale form '${o.tag}`);
+  static ERROR_InvalidLocale = SevError | 0x0002;
+
+  static HardwareLayerHasTooManyRows = () => m(this.ERROR_HardwareLayerHasTooManyRows, `'hardware' layer has too many rows`);
+  static ERROR_HardwareLayerHasTooManyRows = SevError | 0x0003;
 
-export function getErrorSeverityName(code: number): string {
-  let severity = code & CompilerErrorSeverity.Severity_Mask;
-  switch(severity) {
-    case CompilerErrorSeverity.Info: return 'INFO';
-    case CompilerErrorSeverity.Hint: return 'HINT';
-    case CompilerErrorSeverity.Warn: return 'WARN';
-    case CompilerErrorSeverity.Error: return 'ERROR';
-    case CompilerErrorSeverity.Fatal: return 'FATAL';
-    default: return 'UNKNOWN';
+  static RowOnHardwareLayerHasTooManyKeys = (o:{row: number}) =>  m(this.ERROR_RowOnHardwareLayerHasTooManyKeys, `Row #${o.row} on 'hardware' layer has too many keys`);
+  static ERROR_RowOnHardwareLayerHasTooManyKeys = SevError | 0x0004;
+
+  static KeyNotFoundInKeyBag = (o:{keyId: string, col: number, row: number, layer: string, form: string}) =>
+     m(this.ERROR_KeyNotFoundInKeyBag, `Key ${o.keyId} in position #${o.col} on row #${o.row} of layer ${o.layer}, form '${o.form}' not found in key bag`);
+  static ERROR_KeyNotFoundInKeyBag = SevError | 0x0005;
+
+  static OneOrMoreRepeatedLocales = () =>
+    m(this.HINT_OneOrMoreRepeatedLocales, `After minimization, one or more locales is repeated and has been removed`);
+  static HINT_OneOrMoreRepeatedLocales = SevHint | 0x0006;
```

Comment:
Confirming you're intending to include Hints in `CompilerErrors`.
Is it possible later on to also have `SevError | 0x0006;`?

Reply:
1. Yes, all classes of 'error' from info through to fatal. Perhaps we should call it `CompilerMessages`?

2. Each code should be unique independent of its severity code, so we wouldn't use `SevError | 0x0006` in future. (In theory we may want to change the severity of a code one day?)